pcap/pcapio/ngread.go

```diff
@@ -0,0 +1,317 @@
+			return err
+		}
+		b = b[4+len(body):]
+		switch code {
+		case ngOptionCodeInterfaceName:
+			intf.Name = string(body)
+		case ngOptionCodeComment:
+			intf.Comment = string(body)
+		case ngOptionCodeInterfaceDescription:
+			intf.Description = string(body)
+		case ngOptionCodeInterfaceFilter:
+			// ignore filter type (first byte) since it is not specified
+			intf.Filter = string(body[1:])
+		case ngOptionCodeInterfaceOS:
+			intf.OS = string(body)
+		case ngOptionCodeInterfaceTimestampOffset:
+			if len(body) != 8 {
+				return errors.New("bad option value: ngOptionCodeInterfaceTimestampOffset")
+			}
+			intf.TimestampOffset = r.getUint64(body[:8])
+		case ngOptionCodeInterfaceTimestampResolution:
+			if len(body) != 1 {
+				return errors.New("bad option value: ngOptionCodeInterfaceTimestampResolution")
+			}
+			intf.TimestampResolution = NgResolution(body[0])
+		}
+	}
+	if intf.TimestampResolution == 0 {
+		intf.TimestampResolution = 6
+	}
+
+	//parse options
+	if intf.TimestampResolution.Binary() {
+		//negative power of 2
+		intf.secondMask = 1 << intf.TimestampResolution.Exponent()
+	} else {
+		//negative power of 10
+		intf.secondMask = 1
+		for j := uint8(0); j < intf.TimestampResolution.Exponent(); j++ {
+			intf.secondMask *= 10
+		}
+	}
+	intf.scaleDown = 1
+	intf.scaleUp = 1
+	if intf.secondMask < 1e9 {
+		intf.scaleUp = 1e9 / intf.secondMask
+	} else {
+		intf.scaleDown = intf.secondMask / 1e9
+	}
+	//XXX this list should be per section no?
+	r.ifaces = append(r.ifaces, intf)
+	return nil
+}
+
+// convertTime adds offset + shifts the given time value according to the given interface
+func (r *NgReader) convertTime(ifaceID int, ts uint64) (int64, int64) {
+	iface := r.ifaces[ifaceID]
+	return int64(ts/iface.secondMask + iface.TimestampOffset), int64(ts % iface.secondMask * iface.scaleUp / iface.scaleDown)
+}
+
+// readPacketHeader looks for a packet (enhanced, simple, or packet) and parses the header.
+// If an interface descriptor, an interface statistics block, or a section header is encountered, those are handled accordingly.
+// All other block types are skipped. New block types must be added here.
+func (r *NgReader) Read() ([]byte, BlockType, error) {
+	block := r.first
+	if block != nil {
+		r.first = nil
+		r.offset += uint64(len(block))
+		return block, TypeSection, nil
+	}
+	for {
+		typ, block, err := r.readBlock()
+		if err != nil {
+			return nil, 0, err
+		}
+		r.offset += uint64(len(block))
+		switch typ {
+		case ngBlockTypeEnhancedPacket:
+			if len(block) < 28 {
+				return nil, 0, ErrCorruptPcap
+			}
+			ifno := int(r.getUint32(block[8:12]))
+			if ifno >= len(r.ifaces) {
+				return nil, 0, fmt.Errorf("Interface id %d not present in section (have only %d interfaces)", r.ci.InterfaceIndex, len(r.ifaces))
+			}
+			caplen := int(r.getUint32(block[20:24]))
+			//length := int(r.getUint32(block[24:28]))
+			packet := block[28:]
+			if len(packet) < caplen {
+				return nil, 0, ErrCorruptPcap
+			}
```

Comment:
There's a lot of commonality between this and `Packet()`; would it be possible to share the logic?

Reply:
good idea